File: src/main/java/com/numbers/discordbot/audio/TrackScheduler.java
package com.numbers.discordbot.audio;

import com.sedmelluq.discord.lavaplayer.player.*;
import com.sedmelluq.discord.lavaplayer.player.event.*;
import com.sedmelluq.discord.lavaplayer.track.*;
import java.util.concurrent.*;
import java.util.stream.*;

public class TrackScheduler extends AudioEventAdapter {

    private final AudioPlayer player;
    private final BlockingQueue<AudioTrack> queue;

    public TrackScheduler(AudioPlayer player) {
        this.player = player;
        this.queue = new LinkedBlockingQueue<>();
    }

    public Stream<AudioTrack> getQueueStream(){
        return queue.stream();
    }
    
    public long getQueueSize(){
        return queue.size();
    }
    
    public boolean isQueueEmpty(){
        return queue.isEmpty();
    }
    
    public void clear(){
        queue.clear();
    }
    
    public void queue(AudioTrack track) {
        if (!player.startTrack(track, true)) {
            queue.offer(track);
        }
    }

    public void nextTrack() {
        player.startTrack(queue.poll(), false);
    }

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        if (endReason.mayStartNext) {
            nextTrack();
        }
    }
}

File: src/main/java/com/numbers/discordbot/audio/TrackScheduler.kt
package com.numbers.discordbot.audio

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import java.util.*
import java.util.concurrent.LinkedBlockingQueue
import java.util.stream.Stream

class TrackScheduler(private val player: AudioPlayer) : AudioEventAdapter() {

    private val queue : Queue<AudioTrack> = LinkedBlockingQueue<AudioTrack>()

    fun getQueueStream() : Stream<AudioTrack> { return queue.stream() }

    val size : Int get() = queue.size

    val isEmpty : Boolean get() = queue.isEmpty()

    fun clear() = queue::clear

    fun remove(amount : Long ) = (0..amount).forEach { queue.poll() }

    fun queue(track: AudioTrack) {
        if(!player.startTrack(track, true)){
            queue.offer(track)
        }
    }

    fun nextTrack() { player.startTrack(queue.poll(), false) }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        if(endReason.mayStartNext){
            nextTrack()
        }
    }

}
